File: impl/go/project.go
package cson

import "strings"

// reservedPrefix marks the wrapper members; SPEC §6 reserves it so a document
// key can never collide with $value / $confidence / $annotations / $synthesize.
const reservedPrefix = "$"

// Project returns the SPEC §6 JSON projection of the document.
//
// One rule: a node with no metadata projects to its bare value; a node carrying
// confidence or annotations projects to a wrapper object
//
//	{"$value": …, "$confidence": …, "$annotations": [ … ]}
//
// with the metadata members omitted when absent. Because the wrapper is keyed on
// the NODE and not on its parent's key, it applies identically to object values
// and array elements — an array element is a node and may carry metadata.
//
// This keeps absent confidence distinguishable from an explicit ~1.0 (§4.2): the
// $confidence member is present only when stated.
//
// Objects project to map[string]any, so callers can hand the result straight to
// encoding/json. Key order is not preserved by a Go map; the ordered keys live on
// *Object for a printer to use.
func (d *Document) Project() (map[string]any, error) {
	return projectObject(d.Root)
}

func annotationsOf(n *Node) []any {
	out := make([]any, 0, len(n.Annotations))
	for _, a := range n.Annotations {
		var args any
		if a.Args != nil {
			args = *a.Args
		}
		props := map[string]any{}
		for k, v := range a.Properties {
			props[k] = v
		}
		out = append(out, map[string]any{"name": a.Name, "args": args, "properties": props})
	}
	return out
}

// wrap returns the bare value when the node carries no metadata, a wrapper otherwise.
func wrap(n *Node, projected any) any {
	if n.Confidence == nil && len(n.Annotations) == 0 {
		return projected
	}
	out := map[string]any{"$value": projected}
	if n.Confidence != nil {
		out["$confidence"] = *n.Confidence
	}
	if len(n.Annotations) > 0 {
		out["$annotations"] = annotationsOf(n)
	}
	return out
}

func projectValue(v any) (any, error) {
	switch t := v.(type) {
	case *Synthesize:
		// an object, never a bare string, so it cannot be mistaken for data (§4.4)
		return map[string]any{"$synthesize": t.Description}, nil
	case *Object:
		return projectObject(t)
	case []*Node:
		out := make([]any, 0, len(t))
		for _, n := range t {
			pv, err := projectNode(n)
			if err != nil {
				return nil, err
			}
			out = append(out, pv)
		}
		return out, nil
	default:
		return v, nil
	}
}

func projectNode(n *Node) (any, error) {
	pv, err := projectValue(n.Value)
	if err != nil {
		return nil, err
	}
	return wrap(n, pv), nil
}

func projectObject(o *Object) (map[string]any, error) {
	out := make(map[string]any, o.Len())
	for _, k := range o.Keys() {
		if strings.HasPrefix(k, reservedPrefix) {
			return nil, &Error{Reason: "Key \"" + k + "\" uses the reserved '$' prefix"}
		}
		n, _ := o.Get(k)
		pv, err := projectNode(n)
		if err != nil {
			return nil, err
		}
		out[k] = pv
	}
	return out, nil
}

// Loads parses text and returns its SPEC §6 projection.
func Loads(text string) (map[string]any, error) {
	doc, err := Parse(text)
	if err != nil {
		return nil, err
	}
	return doc.Project()
}

File: impl/js/src/project.mjs
/**
 * JSON projection — SPEC.md §6.
 *
 * One rule: a node with no metadata projects to its bare value; a node carrying
 * confidence or annotations projects to a wrapper object
 *
 *     {"$value": …, "$confidence": …, "$annotations": [ … ]}
 *
 * with the metadata members omitted when absent. Because the wrapper is keyed on
 * the NODE and not on its parent's key, it applies identically to object values
 * and array elements — an array element is a node and may carry metadata.
 *
 * This keeps absent confidence distinguishable from an explicit ~1.0 (§4.2): the
 * `$confidence` member is present only when stated.
 */
import { CsonError, Node, Synthesize } from "./model.mjs";

const RESERVED_PREFIX = "$";

function annotationsOf(node) {
  return node.annotations.map((a) => ({
    name: a.name,
    args: a.args,
    properties: { ...a.properties },
  }));
}

/** Bare value when the node carries no metadata, wrapper object otherwise. */
function wrap(node, projected) {
  if (node.confidence === null && node.annotations.length === 0) return projected;
  const out = { $value: projected };
  if (node.confidence !== null) out.$confidence = node.confidence;
  if (node.annotations.length) out.$annotations = annotationsOf(node);
  return out;
}

function projectValue(v) {
  if (v instanceof Synthesize) {
    // an object, never a bare string, so it cannot be mistaken for data (§4.4)
    return { $synthesize: v.description };
  }
  if (v instanceof Map) return projectObject(v);
  if (Array.isArray(v)) return v.map((n) => (n instanceof Node ? projectNode(n) : projectValue(n)));
  return v;
}

function projectNode(node) {
  return wrap(node, projectValue(node.value));
}

function projectObject(map) {
  const out = {};
  for (const [key, node] of map) {
    // §6 reserves $-prefixed keys so a document key can never collide with the
    // wrapper members.
    if (key.startsWith(RESERVED_PREFIX)) {
      throw new CsonError(`Key ${JSON.stringify(key)} uses the reserved '$' prefix`, 0, 0);
    }
    out[key] = node instanceof Node ? projectNode(node) : projectValue(node);
  }
  return out;
}

/** The SPEC §6 projection. */
export function project(root) {
  return projectObject(root);
}

/**
 * Value-only projection: drops all metadata.
 *
 * Not the spec projection — a convenience for callers that want plain data and
 * have no use for confidence or annotations. Lossy by construction.
 */
export function projectValues(root) {
  const val = (v) => {
    if (v instanceof Synthesize) return { $synthesize: v.description };
    if (v instanceof Map) {
      const o = {};
      for (const [k, n] of v) o[k] = n instanceof Node ? val(n.value) : val(n);
      return o;
    }
    if (Array.isArray(v)) return v.map((n) => (n instanceof Node ? val(n.value) : val(n)));
    return v;
  };
  return val(root);
}
